### ISSUE TITLE
Can we rename 'forall intents' to 'task intents'?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#18463]
It's seemed odd for me for some time that we use different terms for
"forall intents" vs. "task intents" given that both concepts share the
same syntax and effectively do similar things.  My inclination would be
to call them both `task intents` since even in a `forall` loop, they are
affecting behavior on a per-task granularity.

On the other hand, if we were to add intents to for loops (https://github.com/chapel-lang/chapel/issues/17857)
and foreach loops (no issue currently) as has been proposed, it could
be that "task intent" isn't the right term given that those are implemented
with a single task?  "loop intent" could work except that we apply them to
`begin` and `cobegin` statements as well.  So maybe we need a completely
new term?

**Comment by mppf** [comment]
I agree that we need to improve the terminology here. FWIW of `task intent` vs `loop intent` I lean more towards `loop intent`. Perhaps `block intent` would cover these cases well and distinguish from formal argument intents.

**Comment by e-kayrakli** [comment]
>  Perhaps block intent would cover these cases well and distinguish from formal argument intents.

What are the differences between task/forall intents and formal intents? We can't have a reduce intent on a formal, but is there anything else? Do we really need to distinguish these two in terms of terminology?

**Comment by mppf** [comment]
Well, we could make an intentional effort to just call them `intents`, but when necessary, clarify we are talking about `formal intent` / `loop intent` / `task intent`. That seems OK to me. If we do that, I'm not so sure that we need a different term for the combination of `loop intent` and `task intent`.

**Comment by mppf** [comment]
> What are the differences between task/forall intents and formal intents? We can't have a reduce intent on a formal, but is there anything else? Do we really need to distinguish these two in terms of terminology?

We also have `var` intent available for task/forall but not for a formal.

**Comment by e-kayrakli** [comment]
Yeah, that is what I was getting at. The differences (`reduce`, `var`, or functionality in general) is not too much, and I think we can just call them "intents".

**Comment by bradcray** [comment]
> What are the differences between task/forall intents and formal intents?
> Do we really need to distinguish these two in terms of terminology?

I agree that they're all intents, but think that being able to distinguish between "argument intents" vs. "return intents" vs. "with-clause intents" (by some name) is useful (if for no other reason than to be able to say which intents are legal in which contexts).

> If we do that, I'm not so sure that we need a different term for the combination of loop intent and task intent.

@mppf: Just to make sure I'm understanding:  Are you essentially suggesting that the intents on a `coforall` or `forall` loop could be called "loop intents" or "task intents" interchangeably?  (whereas `begin` and `cobegin` would always have "task intents" and `for` and `foreach` would always have "loop intents"?).

**Comment by mppf** [comment]
Here is how I would think of it if we choose to go further with `intent` vs adding another term:
 * just `intent` describes all of these and we'd try to use that term most places
 * `coforall` intents are both `task intent` and `loop intent`
 * `begin`/`cobegin` intents are always `task intent`
 * `forall`/`foreach`/`for` intents are `loop intent`

**Comment by e-kayrakli** [comment]
Aren't `forall` intents also both `task intent` and `loop intent` ?

**Comment by mppf** [comment]
Well, a forall could be invoking an iterator that doesn't create tasks, so I'd rather call it a loop intent.

**Comment by mstrout** [comment]
> > What are the differences between task/forall intents and formal intents? We can't have a reduce intent on a formal, but is there anything else? Do we really need to distinguish these two in terms of terminology?

> We also have var intent available for task/forall but not for a formal.

What does a var intent mean for a task/forall?

Why can't we have a reduce intent on a formal?

**Comment by mppf** [comment]
> What does a var intent mean for a task/forall?

It means to create a new variable for every task (and use that within the follower loop running within that task, for a forall). I am expecting that when we get to vectorization & GPU support, it will create a new variable for each vector lane / GPU "thread".

> Why can't we have a reduce intent on a formal?

I don't know what it would mean. Do you?

E.g.

``` chapel
proc f(+ reduce arg) {
  ... something ...
}
var x: int;
f(x); // what does this do?
```

**Comment by mstrout** [comment]
> > Why can't we have a reduce intent on a formal?
> I don't know what it would mean.  Do you?

I was thinking of the function itself as a commutative operation being applied to the formal parameter (https://people.csail.mit.edu/rinard/paper/toplas97.pdf).  Thus signaling that a call to such a function in a forall loop would be able to reorder but not run such function calls concurrently.   Of course, that goes against the `forall` semantics.  In a way similar to diamond tiling in that the loop isn't fully parallel.

Let me turn this around and ask what does a forall/task reduce intent mean?  I see the examples at https://chapel-lang.org/docs/technotes/reduceIntents.html#readme-reduceintents.  For a `forall` there is some number of tasks involved.  What if there is just one task?

```
var x = 0;
forall myIterator() with (+ reduce x) {
  x += 1;
}
```
In the above code, it appears the `reduce` intent is going to modify the semantics of `forall`?  Since it is no longer fully parallel?

**Comment by e-kayrakli** [comment]
> In the above code, it appears the reduce intent is going to modify the semantics of forall? Since it is no longer fully parallel?

I would argue that it is still fully parallel, or at least as parallel as it would have been without the reduce intent. It'll only make the synchronization at the end of the loop do the reduction.

You can think of the `reduce` intent a variation of the `var` intent, in the sense that each task will have its own copy of the variable. And they'll be able to work on their own copies in parallel safely. We just reduce them at the end of the forall as part of the synchronization which exists already.

**Comment by mppf** [comment]
> > > Why can't we have a reduce intent on a formal?
> > I don't know what it would mean.  Do you?
> 
> I was thinking of the function itself as a commutative operation being applied to the formal parameter (https://people.csail.mit.edu/rinard/paper/toplas97.pdf).  Thus signaling that a call to such a function in a forall loop would be able to reorder but not run such function calls concurrently.   Of course, that goes against the `forall` semantics.  In a way similar to diamond tiling in that the loop isn't fully parallel.

I suspect that we're running into a terminology issue here. When I say "formal" I'm talking specifically (only) about the variables defined by a function signature, e.g. `proc f(arg)` has `arg` as a formal. In particular in `var x = 0; forall myIterator() with (+ reduce x)` there is nothing that I would call a "formal". Certainly, `operator +` has "formal"s. Am I right that this is a terminology issue?

Either way, can you show a Chapel example that has a reduce intent applied to a formal?
